README.md
# hipache-api

HTTP interface for [Hipache](https://github.com/hipache/hipache)

## Install

```
go get sosedoff/hipache-api
```

Or use one of the binaries from [Github Releases](https://github.com/sosedoff/hipache-api/releases).

## Usage

Start http server:

```
hipache-api
```

By default hipache-api will try to connect to redis server on `localhost:6379`. 
To change that you can specify extra environment variables:

```
REDIS_HOST=host REDIS_PORT=port hipache-api
```

Enable api key authentication:

```
API_KEY=key hipache-api
```

### Docker

There's a prebuilt Docker image available. First, pull the image:

```
docker pull sosedoff/hipache-api
```

Then start container:

```
docker run -d -p 5000 -e REDIS_HOST=host -e REDIS_PORT=port -e API_KEY=key sosedoff/hipache-api
```

## API

```
   GET /frontends
   GET /frontends/:name
  POST /frontends?host=site.com&backends=http://host1:port,http://host2:port
  POST /frontends/:name?backends=http://host1:port,http://host2:port
DELETE /frontends/:name
DELETE /frontends/:name/backend?backend=http://host1:port
  POST /flush
```

## Build

```
make build   # Make development build
make release # Build binaries for linux/osx
```

## License

The MIT License (MIT)

Copyright (c) 2014 Dan Sosedoff, <dan.sosedoff@gmail.com>

CHANGELOG.md
# 0.2.0 - 2014-12-29

- Fixed internal Redis usage

# 0.1.0 - 2014-12-21

Initial release

hipache.go
package main

import (
	"fmt"
	"strings"

	"github.com/garyburd/redigo/redis"
)

type Hipache struct {
	Redis redis.Conn
}

func NewHipache(host string) (Hipache, error) {
	conn, err := redis.Dial("tcp", host)
	return Hipache{conn}, err
}

func (h Hipache) Close() {
	h.Redis.Close()
}

func (h Hipache) Frontends() ([]string, error) {
	keys, err := redis.Strings(h.Redis.Do("KEYS", "frontend:*"))

	if err != nil {
		return []string{}, err
	}

	for i, val := range keys {
		keys[i] = strings.Replace(val, "frontend:", "", 1)
	}

	return keys, nil
}

func (h Hipache) Backends(frontend string) ([]string, error) {
	values, err := redis.Strings(h.Redis.Do("LRANGE", "frontend:"+frontend, "1", "-1"))

	if err != nil {
		return []string{}, err
	}

	return values, nil
}

func (h Hipache) AddFrontend(host string) error {
	_, err := h.Redis.Do("RPUSH", "frontend:"+host, host)
	return err
}

func (h Hipache) RemoveFrontend(frontend string) error {
	_, err := h.Redis.Do("DEL", "frontend:"+frontend)
	return err
}

func (h Hipache) AddBackend(frontend string, backend string) error {
	_, err := h.Redis.Do("RPUSH", "frontend:"+frontend, backend)
	return err
}

func (h Hipache) RemoveBackend(frontend string, backend string) error {
	_, err := h.Redis.Do("LREM", "frontend:"+frontend, "0", backend)
	return err
}

func (h Hipache) FrontendExists(frontend string) (bool, error) {
	return redis.Bool(h.Redis.Do("EXISTS", "frontend:"+frontend))
}

func (h Hipache) Flush() error {
	frontends, err := h.Frontends()

	if err != nil {
		return err
	}

	for _, fe := range frontends {
		_, err = h.Redis.Do("DEL", "frontend:"+fe)

		if err != nil {
			fmt.Println("ERROR:", err)
		}
	}

	return nil
}


endpoints.go
package main

import (
	"strings"

	"github.com/gin-gonic/gin"
)

type Err struct {
	Status  int    `json:"status"`
	Message string `json:"error"`
}

func ErrorResponse(code int, message string, c *gin.Context) {
	c.JSON(code, Err{code, message})
}

func SuccessResponse(message string, c *gin.Context) {
	c.JSON(200, map[string]string{"message": message})
}

func GetFrontends(c *gin.Context) {
	frontends, err := hipache.Frontends()

	if err != nil {
		ErrorResponse(400, err.Error(), c)
		return
	}

	c.JSON(200, frontends)
}

func GetBackends(c *gin.Context) {
	fe := c.Params.ByName("fe")
	exists, err := hipache.FrontendExists(fe)

	if err != nil {
		ErrorResponse(400, err.Error(), c)
		return
	}

	if !exists {
		ErrorResponse(404, "Frontend does not exist", c)
		return
	}

	backends, err := hipache.Backends(c.Params.ByName("fe"))

	if err != nil {
		ErrorResponse(400, err.Error(), c)
		return
	}

	c.JSON(200, backends)
}

func DeleteFrontend(c *gin.Context) {
	fe := c.Params.ByName("fe")
	exists, err := hipache.FrontendExists(fe)

	if err != nil {
		ErrorResponse(400, err.Error(), c)
		return
	}

	if !exists {
		ErrorResponse(404, "Frontend does not exist", c)
		return
	}

	err = hipache.RemoveFrontend(fe)

	if err != nil {
		ErrorResponse(400, err.Error(), c)
		return
	}

	SuccessResponse("Frontend has been deleted", c)
}

func FlushFrontends(c *gin.Context) {
	err := hipache.Flush()

	if err != nil {
		ErrorResponse(400, err.Error(), c)
		return
	}

	c.JSON(200, map[string]string{"message": "All frontends are now flushed"})
}

func CreateFrontend(c *gin.Context) {
	host := c.Request.FormValue("host")
	backends := c.Request.FormValue("backends")

	if host == "" {
		ErrorResponse(400, "Missing parameter 'host'", c)
		return
	}

	exists, _ := hipache.FrontendExists(host)

	if exists {
		ErrorResponse(400, "Frontend already exists", c)
		return
	}

	err := hipache.AddFrontend(host)

	if err != nil {
		ErrorResponse(400, err.Error(), c)
		return
	}

	if backends != "" {
		for _, be := range strings.Split(backends, ",") {
			// TODO: Some error checking!
			hipache.AddBackend(host, be)
		}
	}

	SuccessResponse("Frontend has been created", c)
}

func CreateBackend(c *gin.Context) {
	fe := c.Params.ByName("fe")
	exists, _ := hipache.FrontendExists(fe)

	if !exists {
		ErrorResponse(404, "Frontend does not exist", c)
		return
	}

	backends := c.Request.FormValue("backends")

	if backends == "" {
		ErrorResponse(400, "Missing 'backends' parameter", c)
		return
	}

	for _, be := range strings.Split(backends, ",") {
		hipache.AddBackend(fe, be)
	}

	SuccessResponse("Backends created", c)
}

func DeleteBackend(c *gin.Context) {
	fe := c.Params.ByName("fe")
	be := c.Request.FormValue("backend")

	if be == "" {
		ErrorResponse(400, "Missing 'backend' parameter", c)
		return
	}

	exists, _ := hipache.FrontendExists(fe)
	if !exists {
		ErrorResponse(400, "Frontend does not exist", c)
		return
	}

	err := hipache.RemoveBackend(fe, be)

	if err != nil {
		ErrorResponse(400, err.Error(), c)
		return
	}

	SuccessResponse("Backend removed", c)
}


main.go
package main

import (
	"fmt"
	"os"

	"github.com/gin-gonic/gin"
)

const VERSION = "0.2.0"

var hipache Hipache
var service *gin.Engine

func getEnv(name string, def string) string {
	val := os.Getenv(name)

	if val == "" {
		return def
	}

	return val
}

func setupHipache() {
	host := fmt.Sprintf(
		"%s:%s",
		getEnv("REDIS_HOST", "localhost"),
		getEnv("REDIS_PORT", "6379"),
	)

	conn, err := NewHipache(host)

	if err != nil {
		fmt.Println("Redis connection error:", err)
		os.Exit(1)
	}

	hipache = conn
}

func setupMiddleware() {
	if os.Getenv("API_KEY") != "" {
		service.Use(RequireAuth)
	}
	service.Use(RequireHipache)
}

func setupEndpoints() {
	service.GET("/frontends", GetFrontends)
	service.POST("/frontends", CreateFrontend)
	service.GET("/frontends/:fe", GetBackends)
	service.POST("/frontends/:fe", CreateBackend)
	service.DELETE("/frontends/:fe", DeleteFrontend)
	service.DELETE("/frontends/:fe/backend", DeleteBackend)
	service.POST("/flush", FlushFrontends)
}

func main() {
	// Do not print extra debugging information
	gin.SetMode("release")

	service = gin.Default()

	setupHipache()
	setupMiddleware()
	setupEndpoints()

	defer hipache.Close()

	port := getEnv("PORT", "5000")
	fmt.Printf("Starting hipache-api v%v on 0.0.0.0:%s\n", VERSION, port)
	service.Run(":" + port)
}
